lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            _configurationProvider = new Dictionary<string, Tuple<IConfiguration, IConfiguration>>();
+            _failedEvaluators = new HashSet<string>();
+            _submittedEvaluators = new HashSet<string>();
+            _numNodes = numNodes;
+            _groupCommDriver = groupCommDriver;
+            _configurationManager = configurationManager;
+            _tcpPortProviderConfig = tcpPortProviderConfig;
+            _assignedPartitionDescriptors = 0;
+            _partitionDescriptors = partitionDescriptors;
+            _lock = new object();
+        }
+
+        /// <summary>
+        /// Handles failed evaluator. Moves the id from 
+        /// submitted evaluator to failed evaluator
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        internal void EvaluatorFailed(string evaluatorId)
+        {
+            lock (_lock)
+            {
+                if (!_submittedEvaluators.Contains(evaluatorId))
+                {
+                    Exceptions.Throw(new Exception("Failed evaluator was never submitted"), Logger);
+                }
+
+                _failedEvaluators.Add(evaluatorId);
+                _submittedEvaluators.Remove(evaluatorId);
+            }
+        }
+
+        /// <summary>
+        /// Gives configuration for next evaluator either from failed 
+        /// evaluator or new configuration
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        /// <returns></returns>
+        internal Tuple<IConfiguration, IConfiguration> GetNextConfiguration(string evaluatorId)
+        {
+            lock (_lock)
+            {
+                if (_submittedEvaluators.Contains(evaluatorId))
+                {
+                    Exceptions.Throw(new Exception("The evaluator is already submitted"), Logger);
+                }
+
+                if (_failedEvaluators.Count == 0 && _assignedPartitionDescriptors >= _numNodes)
+                {
+                    Exceptions.Throw(new Exception("No more configuration can be provided"), Logger);
+                }
+
+                //If some failed id exists return that configuration
+                if (_failedEvaluators.Count != 0)
+                {
+                    string failedEvaluatorId = _failedEvaluators.First();
+                    _failedEvaluators.Remove(failedEvaluatorId);
+                    _configurationProvider[evaluatorId] = _configurationProvider[failedEvaluatorId];
+                }
+                else
+                {
+                    _assignedPartitionDescriptors++;
+
+                    //Checks whether to put update task configuration or map task configuration
+                    if (_assignedPartitionDescriptors == 1)
+                    {
+                        _configurationProvider[evaluatorId] = GetUpdateTaskContextAndServiceConfiguration();
```

Comment:
Is it possible..I mean when we get the new evaluators, is there a possibility that Ids can be same. Can you comment on this?